File: backend/app/services/billing_calculations.py
"""JMD billing calculations and ready-for-pickup bill publishing."""

from decimal import Decimal

from app.data.revised_rate_table import MAX_AUTO_RATE_LBS, QUOTE_MESSAGE
from app.models.package import Package
from app.services.shipping_service import billable_weight_lbs, calculate_shipping_cost


def compute_total_due(
    freight: Decimal | None,
    duties: Decimal | None,
    handling: Decimal | None,
    other: Decimal | None,
) -> Decimal | None:
    parts = [freight, duties, handling, other]
    if all(p is None for p in parts):
        return None
    total = Decimal("0")
    for part in parts:
        if part is not None:
            total += part
    return total.quantize(Decimal("0.01"))


def _decimal(value) -> Decimal | None:
    if value is None:
        return None
    return Decimal(str(value)).quantize(Decimal("0.01"))


def ensure_freight_jmd(package: Package) -> None:
    """Set shipping charge from billable weight when releasing from customs."""
    if package.billable_weight_lbs is None and package.actual_weight_lbs is None:
        raise ValueError(f"{package.tracking_number} has no weight on file")

    if package.estimated_freight_jmd is not None and package.estimated_freight_jmd > 0:
        return

    weight = package.actual_weight_lbs or package.billable_weight_lbs
    billable = billable_weight_lbs(weight)
    if billable > MAX_AUTO_RATE_LBS:
        raise ValueError(f"{package.tracking_number}: {QUOTE_MESSAGE}")

    quote = calculate_shipping_cost(weight)
    package.estimated_freight_jmd = _decimal(quote["cost_jmd"])
    package.rate_tier_label = quote["tier_label"]
    if package.billable_weight_lbs is None:
        package.billable_weight_lbs = quote["billable_weight_lbs"]


def publish_ready_for_pickup_bill(
    package: Package,
    *,
    estimated_freight_jmd: float | None = None,
    duties_jmd: float | None = None,
    handling_jmd: float | None = None,
    other_fees_jmd: float | None = None,
) -> None:
    if estimated_freight_jmd is not None:
        package.estimated_freight_jmd = _decimal(estimated_freight_jmd)
    ensure_freight_jmd(package)

    if duties_jmd is not None:
        package.duties_jmd = _decimal(duties_jmd)
    if handling_jmd is not None:
        package.handling_jmd = _decimal(handling_jmd)
    if other_fees_jmd is not None:
        package.other_fees_jmd = _decimal(other_fees_jmd)

    package.total_due_jmd = compute_total_due(
        package.estimated_freight_jmd,
        package.duties_jmd,
        package.handling_jmd,
        package.other_fees_jmd,
    )
    if package.total_due_jmd is None or package.total_due_jmd <= 0:
        raise ValueError(f"Could not calculate bill for {package.tracking_number}")
    package.billing_status = "ready"

File: server/src/services/billingCalculations.ts
import { MAX_AUTO_RATE_LBS, QUOTE_MESSAGE } from '../data/revisedRateTable.js'
import type { packages } from '../db/schema/index.js'
import { billableWeightLbs, calculateShippingCost } from './shippingService.js'

type PackageRow = typeof packages.$inferSelect

function decimal(value: number | string | null | undefined): number | null {
  if (value == null) return null
  const n = typeof value === 'number' ? value : parseFloat(String(value))
  if (Number.isNaN(n)) return null
  return Math.round(n * 100) / 100
}

export function computeTotalDue(
  freight: number | null,
  duties: number | null,
  handling: number | null,
  other: number | null,
): number | null {
  if ([freight, duties, handling, other].every((p) => p == null)) return null
  let total = 0
  for (const part of [freight, duties, handling, other]) {
    if (part != null) total += part
  }
  return Math.round(total * 100) / 100
}

export function ensureFreightJmd(pkg: PackageRow): void {
  if (pkg.billableWeightLbs == null && pkg.actualWeightLbs == null) {
    throw new Error(`${pkg.trackingNumber} has no weight on file`)
  }

  const existingFreight = pkg.estimatedFreightJmd != null ? parseFloat(pkg.estimatedFreightJmd) : null
  if (existingFreight != null && existingFreight > 0) return

  const weight =
    pkg.actualWeightLbs != null
      ? parseFloat(pkg.actualWeightLbs)
      : pkg.billableWeightLbs ?? 0
  const billable = billableWeightLbs(weight)
  if (billable > MAX_AUTO_RATE_LBS) {
    throw new Error(`${pkg.trackingNumber}: ${QUOTE_MESSAGE}`)
  }

  const quote = calculateShippingCost(weight)
  ;(pkg as { estimatedFreightJmd?: string | null }).estimatedFreightJmd = String(quote.cost_jmd)
  ;(pkg as { rateTierLabel?: string | null }).rateTierLabel = quote.tier_label
  if (pkg.billableWeightLbs == null) {
    ;(pkg as { billableWeightLbs?: number | null }).billableWeightLbs = quote.billable_weight_lbs
  }
}

export function publishReadyForPickupBill(
  pkg: PackageRow,
  options?: {
    estimatedFreightJmd?: number | null
    dutiesJmd?: number | null
    handlingJmd?: number | null
    otherFeesJmd?: number | null
  },
): void {
  if (options?.estimatedFreightJmd != null) {
    ;(pkg as { estimatedFreightJmd?: string | null }).estimatedFreightJmd = String(
      decimal(options.estimatedFreightJmd),
    )
  }
  ensureFreightJmd(pkg)

  if (options?.dutiesJmd != null) {
    ;(pkg as { dutiesJmd?: string | null }).dutiesJmd = String(decimal(options.dutiesJmd))
  }
  if (options?.handlingJmd != null) {
    ;(pkg as { handlingJmd?: string | null }).handlingJmd = String(decimal(options.handlingJmd))
  }
  if (options?.otherFeesJmd != null) {
    ;(pkg as { otherFeesJmd?: string | null }).otherFeesJmd = String(decimal(options.otherFeesJmd))
  }

  const freight = pkg.estimatedFreightJmd != null ? parseFloat(pkg.estimatedFreightJmd) : null
  const duties = pkg.dutiesJmd != null ? parseFloat(pkg.dutiesJmd) : null
  const handling = pkg.handlingJmd != null ? parseFloat(pkg.handlingJmd) : null
  const other = pkg.otherFeesJmd != null ? parseFloat(pkg.otherFeesJmd) : null
  const total = computeTotalDue(freight, duties, handling, other)

  if (total == null || total <= 0) {
    throw new Error(`Could not calculate bill for ${pkg.trackingNumber}`)
  }

  ;(pkg as { totalDueJmd?: string | null }).totalDueJmd = String(total)
  ;(pkg as { billingStatus?: string }).billingStatus = 'ready'
}
